storage/indexer/src/db_indexer.rs

```diff
@@ -0,0 +1,364 @@
+            committer_handle: Some(committer_handle),
+        }
+    }
+
+    pub fn get_persisted_version(&self) -> Version {
+        // read the latest key from the db
+        let mut rev_iter_res = self
+            .db
+            .rev_iter::<TailerMetadataSchema>()
+            .expect("Cannot create db tailer metadata iterator");
+        rev_iter_res.seek_to_last();
+        rev_iter_res
+            .next()
+            .map(|res| res.map_or(0, |(version, _)| version))
+            .unwrap_or_default()
+    }
+
+    pub fn event_enabled(&self) -> bool {
+        self.config.enable_event
+    }
+
+    pub fn transaction_enabled(&self) -> bool {
+        self.config.enable_transaction
+    }
+
+    pub fn process_a_batch(&self, start_version: Option<Version>) -> Result<Version> {
+        let mut version = start_version.unwrap_or(0);
+        let db_iter: Box<
+            dyn Iterator<
+                Item = std::prelude::v1::Result<
+                    (aptos_types::transaction::Transaction, Vec<ContractEvent>),
+                    AptosDbError,
+                >,
+            >,
+        > = self
+            .main_db_reader
+            .get_db_tailor_iter(version, self.config.batch_size)?;
+        let batch = SchemaBatch::new();
+        db_iter.for_each(|res| {
+            res.map(|(txn, events)| {
+                if let Some(txn) = txn.try_as_signed_user_txn() {
+                    if self.config.enable_transaction {
+                        batch
+                            .put::<TransactionByAccountSchema>(
+                                &(txn.sender(), txn.sequence_number()),
+                                &version,
+                            )
+                            .expect("Failed to put transactions to db tailer batch");
+                    }
+
+                    if self.config.enable_event {
+                        events.iter().enumerate().for_each(|(idx, event)| {
+                            if let ContractEvent::V1(v1) = event {
+                                batch
+                                    .put::<EventByKeySchema>(
+                                        &(*v1.key(), v1.sequence_number()),
+                                        &(version, idx as u64),
+                                    )
+                                    .expect("Failed to events by key to db tailer batch");
+                                batch
+                                    .put::<EventByVersionSchema>(
+                                        &(*v1.key(), version, v1.sequence_number()),
+                                        &(idx as u64),
+                                    )
+                                    .expect("Failed to events by version to db tailer batch");
+                            }
+                        });
+                    }
+                }
+                version += 1;
+            })
+            .expect("Failed to iterate db tailer iterator");
```

Comment:
just `let (txn, events) = res?` in the beginning of the for_each is fine? -- reduces one level of indentation.

Reply:
cannot return error in for_each. added let (txn, events) = res.expect("xxxx");